[Docs] Downloader cog guide (#4511)
Cog-Creators/Red-DiscordBot: docs/cog_guides/downloader.rst

Install community cogs made by Cog Creators.

Community cogs, also called third party cogs, are not included
in the default Red install.

Community cogs come in repositories. Repos are a group of cogs
you can install. You always need to add the creator's repository
using the ``[p]repo`` command before you can install one or more
cogs from the creator.

.. _downloader-commands:

--------
Commands
--------

.. _downloader-command-cog:

^^^
cog
^^^

.. note:: |owner-lock|

**Syntax**

.. code-block:: none

    [p]cog 

**Description**

Base command for cog installation management commands.

.. _downloader-command-cog-checkforupdates:

"""""""""""""""""""
cog checkforupdates
"""""""""""""""""""

**Syntax**

.. code-block:: none

    [p]cog checkforupdates 

**Description**

Check for available cog updates (including pinned cogs).

This command doesn't update cogs, it only checks for updates.
Use ``[p]cog update`` to update cogs.

.. _downloader-command-cog-info:

""""""""
cog info
""""""""

**Syntax**

.. code-block:: none

    [p]cog info <repo_name> <cog_name>

**Description**

List information about a single cog.

Example:
    - ``[p]cog info 26-Cogs defender``

**Arguments**

> - ``<repo_name>`` The repo to get cog info from..

Reviewer: ```suggestion
- ``<repo_name>`` The repo to get cog info from.
```

- ``<cog_name>`` The cog to get info on.

.. _downloader-command-cog-install:

"""""""""""
cog install
"""""""""""

**Syntax**

.. code-block:: none

    [p]cog install <repo_name> <cogs>

**Description**

Install a cog from the given repo.

Examples:
    - ``[p]cog install 26-Cogs defender``
    - ``[p]cog install Laggrons-Dumb-Cogs say roleinvite``

**Arguments**

- ``<repo_name>`` The name of the repo to install cogs from.
- ``<cogs>`` The cog or cogs to install.

.. _downloader-command-cog-installversion:

""""""""""""""""""
cog installversion
""""""""""""""""""

**Syntax**

.. code-block:: none

    [p]cog installversion <repo_name> <revision> <cogs>

**Description**

Install a cog from the specified revision of given repo.

Example:
    - ``[p]cog installversion Broken-Repo e798cc268e199612b1316a3d1f193da0770c7016 cog_name``

**Arguments**

- ``<repo_name>`` The name of the repo to install cogs from.

> - ``<revision>`` The revision to install from.

Reviewer: Expanding what a "revision" is could be useful since end users probably don't know much about git.

Author: Added an explanation of revision with a link and an image. It still makes assumptions that the user know what they're doing though.

- ``<cogs>`` The cog or cogs to install.

.. _downloader-command-cog-list:

""""""""
cog list
""""""""

**Syntax**

.. code-block:: none

    [p]cog list <repo_name>

**Description**

List all available cogs from a single repo.

Example:
    - ``[p]cog list 26-Cogs``

**Arguments**

- ``<repo_name>`` The repo to list cogs from.

.. _downloader-command-cog-listpinned:

""""""""""""""
cog listpinned
""""""""""""""

**Syntax**

.. code-block:: none

    [p]cog listpinned 

**Description**

List currently pinned cogs.

.. _downloader-command-cog-pin:

"""""""
cog pin
"""""""

**Syntax**

.. code-block:: none

    [p]cog pin <cogs>

**Description**

Pin cogs - this will lock cogs on their current version.

Examples:
    - ``[p]cog pin defender``
    - ``[p]cog pin outdated_cog1 outdated_cog2``

**Arguments**

- ``<cogs>`` The cog or cogs to pin. Must already be installed.

.. _downloader-command-cog-reinstallreqs:

"""""""""""""""""
cog reinstallreqs
"""""""""""""""""

**Syntax**

.. code-block:: none

    [p]cog reinstallreqs 

**Description**

This command will reinstall cog requirements and shared libraries for all installed cogs.

> Red might ask user to use this when it clears contents of lib folder

Reviewer: ```suggestion
Red might ask the owner to use this when it clears contents of the lib folder
```

Reviewer: While I do think this grammar change should be done in the docstring, this command should **not** be included in the docs  as it is marked as hidden.

Reviewer 2: The command is not actually hidden, I've mistakenly told Flame that it is.
Nonetheless, I still don't think the command should really be shown in the user guide as it's generally only supposed to be used if Red asks you to and it's not a command that the user really needs to learn about from this guide.

Author: Would it be more correct to make the command hidden? I will remove from the guide regardless.

Author: Also added a comment that this command shouldn't be used unless Red asks for it.

Reviewer 2: Seems fine to me to hide it.

Author: Added PR to hide this in #4590

> because of change in minor version of Python.

Reviewer: ```suggestion
because of a change in the minor version of Python.
```

.. _downloader-command-cog-uninstall:

"""""""""""""
cog uninstall
"""""""""""""

**Syntax**

.. code-block:: none

    [p]cog uninstall <cogs>

**Description**

Uninstall cogs.

You may only uninstall cogs which were previously installed
by Downloader.

Examples:
    - ``[p]cog uninstall 26-Cogs defender``
    - ``[p]cog uninstall Laggrons-Dumb-Cogs say roleinvite``

**Arguments**

- ``<cogs>`` The cog or cogs to uninstall.

.. _downloader-command-cog-unpin:

"""""""""
cog unpin
"""""""""

**Syntax**

.. code-block:: none

    [p]cog unpin <cogs>

**Description**

> Unpin cogs - this will remove update lock from cogs.

Reviewer: ```suggestion
Unpin cogs - this will remove the update lock from those cogs.
```

Examples:
    - ``[p]cog unpin defender``
    - ``[p]cog unpin updated_cog1 updated_cog2``

**Arguments**

- ``<cogs>`` The cog or cogs to unpin. Must already be installed and pinned.

.. _downloader-command-cog-update:

""""""""""
cog update
""""""""""

**Syntax**

.. code-block:: none

    [p]cog update [cogs...]

**Description**

Update all cogs, or ones of your choosing.

Examples:
    - ``[p]cog update``
    - ``[p]cog update defender``

**Arguments**

> - ``<cogs>`` The cog or cogs to update. If omitted, all cogs are updated.

Reviewer: Should this be using square brackets? Is this a mistake of the auto-generator?

Author: Swapped to square brackets and reviewed all other bracket to make sure the match the `usage`

.. _downloader-command-cog-updateallfromrepos:

""""""""""""""""""""""
cog updateallfromrepos
""""""""""""""""""""""

**Syntax**

.. code-block:: none

    [p]cog updateallfromrepos <repos>

**Description**

Update all cogs from repos of your choosing.

Examples:
    - ``[p]cog updateallfromrepos 26-Cogs``
    - ``[p]cog updateallfromrepos Laggrons-Dumb-Cogs 26-Cogs``

**Arguments**

- ``<repos>`` The repo or repos to update all cogs from.

.. _downloader-command-cog-updatetoversion:

"""""""""""""""""""
cog updatetoversion
"""""""""""""""""""

**Syntax**

.. code-block:: none

    [p]cog updatetoversion <repo_name> <revision> [cogs]

**Description**

Update all cogs, or ones of your choosing to chosen revision of one repo.

> Note that update doesn't mean downgrade and therefore revision
> has to be newer than the one that cog currently has. If you want to
> downgrade the cog, uninstall and install it again.

Reviewer: ```suggestion
Note that update doesn't mean downgrade and therefore `revision`
has to be newer than the version that cog currently has installed. If you want to
downgrade the cog, uninstall and install it again.
```

Example:
    - ``[p]cog updatetoversion Broken-Repo e798cc268e199612b1316a3d1f193da0770c7016 cog_name``

**Arguments**

- ``<repo_name>`` The repo or repos to update all cogs from.

> - ``<revision>`` The revision to update to.

Reviewer: Same note about "revision" as above.

Author: Added a comment pointing to `cog installversion` to get an explanation of revision

The branch will be the default branch if not specified.

**Arguments**

- ``<name>`` The name given to the repo.
- ``<repo_url>`` URL to the cog branch. Usually GitHub or GitLab.
- ``[branch]`` Optional branch to install cogs from.

.. _downloader-command-repo-delete:

"""""""""""
repo delete
"""""""""""

**Syntax**

.. code-block:: none

    [p]repo delete <repo_name>

.. tip:: Aliases: ``repo remove``, ``repo del``

**Description**

Remove a repo and its files.

Example:
    - ``[p]repo delete 26-Cogs``

**Arguments**

- ``<repo_name>`` The name of an already added repo

.. _downloader-command-repo-info:

"""""""""
repo info
"""""""""

**Syntax**

.. code-block:: none

    [p]repo info <repo_name>

**Description**

Show information about a repo.

Example:
    - ``[p]repo info 26-Cogs``

**Arguments**

- ``<repo_name>`` The name of the repo to show info about.

.. _downloader-command-repo-list:

"""""""""
repo list
"""""""""

**Syntax**

.. code-block:: none

    [p]repo list 

**Description**

List all installed repos.

.. _downloader-command-repo-update:

"""""""""""
repo update
"""""""""""

**Syntax**

.. code-block:: none

    [p]repo update [repos...]

**Description**

> Update all repos, or ones of your choosing.

Reviewer: Maybe add a note that this does **not** update the installed cogs?

Author: Comment added below this line
